### ISSUE TITLE
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xaa in position 196607: invalid start byte

### ISSUE THREAD

**Issue by DanielEbert** [justfoxing/ghidra_bridge#66]
Hello,

I want to remotify a function 'my_function' and return a very large object from my_function.
When I do this, I get the following output:
~~~
2022-02-17 12:04:47,842 [ERROR bridge.py:522 run()] 'utf-8' codec can't decode byte 0xaa in position 196607: invalid start byte
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/jfx_bridge/bridge.py", line 504, in run
    msg_dict = json.loads(data.decode("utf-8"))
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xaa in position 196607: invalid start byte
~~~

In the bridge.py file in function read_size_and_data_from_socket, I think the size of 'data' is stored in 32 bits. Maybe there is a problem if 'my_function' returns an object that is larger than 2**32 bytes (~4 Gigabytes)?

Let me know if you need more info.

**Comment by justfoxing** [comment]
Yeah, there's a 2^32 limit on how much can be sent in a message - I didn't think people would be shipping more than 4 gb (and if you are, you may want to try to identify alternatives to doing that - ghidra_bridge is definitely not going to perform well under those conditions, in either memory or network speed).

But! I don't think that's the problem here. The error message is having trouble decoding the received JSON, with the decode fail happening at 196607 - nowhere near a 2^32 limit (but still a way big message - at least 191Kb :o ). Tracking down invalid unicode always sucks - it'd help to be able to see the code for your "my_function".

**Comment by maxeisele** [comment]
Indeed, that is a lot of data that get's serialized to JSON and definitively not the most efficient way. What's actually done is passing edges from the control flow graph of the program. However, the error does only occur once in a while, but is is always `byte 0xaa` at different positions. If I find a way to reproduce it deterministically, I will let you know.

What's your recommendation on passing larger data back to python? Maybe a pipe or so?

**Comment by justfoxing** [comment]
Honestly, if you're not working across the network, I'd probably suggest
writing it straight out to a file and reading it back from disk in the
python process. Take ghidra_bridge and any networking complexity right out
of the picture. Also means that when something goes a little off, you can
maybe resume from the file instead of having to collect the data from
scratch.

**Comment by justfoxing** [comment]
Additionally, if you want, you could try patching the jfx_bridge/bridge.py on the receiving side to log the message when it hits the decode issue.

This would look something like replacing the line `msg_dict = json.loads(data.decode("utf-8"))` in BridgeReceiverThread.run() with something like the following:

```
                try:
                    msg_dict = json.loads(data.decode("utf-8"))
                except Exception:
                    with open("bad_message.bin", "wb") as output:
                        output.write(data)
                    raise
```

**Comment by justfoxing** [comment]
Hah! Just enough for me to see what's going on. Looks like halfway through a message being received, a second message is jumping in. The unicode error is being caused by the \x00\x00\x00\xaa bytes of the second message size, and there's 0xaa bytes of second message JSON before the initial message resumes. I'm going to have to go hunt through the network dispatching code to see why that can happen...
